[dagster-airflow] from airflow to dagster guide (#10923)
dagster-io/dagster: docs/content/integrations/airflow/from-airflow-to-dagster.mdx

---
title: "Learning Dagster from Airlfow"
description: How to get started with Dagster from an Airflow background
---

# Learning Dagster from Airflow

> In this tutorial, we'll help you make the switch from Airflow to Dagster. Here, we review an Airflow DAG and show how the same functionality and be achieved in Dagster.

Reviewer: This should be `and show how the same functionality can be achieved in Dagster.`

---

> ## Comparing an Airflow Dag to Dagster

Reviewer: `## Comparing an Airflow DAG to Dagster`

> In this guide we will rewrite an [Airflow DAG](https://airflow.apache.org/docs/apache-airflow/stable/tutorial/fundamentals.html) as a Dagster [Job](/concepts/ops-jobs-graphs/jobs). Starting with a basic Airflow DAG:

Reviewer: > In this tutorial, we'll rewrite an [Airflow DAG](https://airflow.apache.org/docs/apache-airflow/stable/tutorial/fundamentals.html) as a Dagster [job](/concepts/ops-jobs-graphs/jobs). Let's start with a basic Airflow DAG:

### Step 1: Defining the Ops

In Dagster, the minimum unit of computation is an op. This directly corresponds to an operator in Airflow. Here, we map the Airflow operators `t1`, `t2`, and `t3` to their respective Dagster op's.

```python file=/integrations/airflow/tutorial_rewrite_dagster.py startafter=start_ops endbefore=end_ops
@op
def print_date(context) -> datetime:
    ds = datetime.now()
    context.log.info(ds)
    return ds

@op(retry_policy=RetryPolicy(max_retries=3), ins={"start": In(Nothing)})
def sleep():
    time.sleep(5)

@op
def templated(context, ds: datetime):
    for _i in range(5):
        context.log.info(ds)
        context.log.info(ds - timedelta(days=7))
```

this yields us the following graph of computations

<center>
  <Image
    alt="Screenshot of the dagster UI, showing the newly created graph of tutorial Ops"
    src="/images/integrations/airflow/airflow_tutorial_rewrite_ops.png"
    width={2200}
    height={1300}
  />
</center>

##### Op Level Retries

In the tutorial DAG, the `t2` operator allowed for 3 retries. To configure the same behavior in Dagster, you can use [op level retry policies](/concepts/ops-jobs-graphs/op-retries)

> ### Step 2: Defining the Job

Reviewer: `## Step 2: Define the job`

> In Dagster, the computations defined in `op`s are composed in `job`s, which define the sequence and dependency structure of the computations you want to execute. This directly corresponds to a DAG in Airflow. Here, we compose the `op`s `print_date`, `sleep` and `templated` to match the dependency structure defined by the Airflow operators `t1`, `t2`, and `t3`.\`\`\`

Reviewer 2: ```suggestion
In Dagster, the computations defined in `op`s are composed in `job`s, which define the sequence and dependency structure of the computations you want to execute. This directly corresponds to a DAG in Airflow. Here, we compose the `op`s `print_date`, `sleep` and `templated` to match the dependency structure defined by the Airflow operators `t1`, `t2`, and `t3`.
```

> In Dagster, the computations defined in `op`s are composed in `job`s, which define the sequence and dependency structure of the computations you want to execute. This directly corresponds to a DAG in Airflow. Here, we compose the `op`s `print_date`, `sleep` and `templated` to match the dependency structure defined by the Airflow operators `t1`, `t2`, and `t3`.

Reviewer: > In Dagster, the computations defined in ops are composed in jobs, which define the sequence and dependency structure of the computations you want to execute. This directly corresponds to a DAG in Airflow. Here, we compose the ops `print_date`, `sleep` and `templated` to match the dependency structure defined by the Airflow operators `t1`, `t2`, and `t3`.

No need to put ops and jobs in code format.

```python file=/integrations/airflow/tutorial_rewrite_dagster.py startafter=start_job endbefore=end_job
@job(tags={"dagster/max_retries": 1, "dag_name": "example"})
def tutorial_job():
    ds = print_date()
    sleep(ds)
    templated(ds)
```

> ##### Job Level Retries

Reviewer: `### Job-level retries`

Headings should use sentence casing.

> Job level retries are managed by the [run launcher](/deployment/run-retries) you will need to enable support for them in your dagster.yaml, once you do you can define the retry count on the job.

Reviewer: > Job-level retries are managed by the [run launcher](/deployment/run-retries). Once enabled in your `dagster.yaml` file, you can define the retry count for the job.

> ### Step 3: Defining the schedule

Reviewer: ` ## Step 3: Define the schedule`

> In Dagster Schedules are defined seperately from your computations, Below we define a schedule that will run the `tutorial_job` daily:

Reviewer 2: lowercase 
```suggestion
In Dagster, schedules can be defined for jobs, which determine the cadence at which a job is triggered to be executed. Below we define a schedule that will run the `tutorial_job` daily:
```

```python file=/integrations/airflow/tutorial_rewrite_dagster.py startafter=start_schedule endbefore=end_schedule
schedule = ScheduleDefinition(job=tutorial_job, cron_schedule="@daily")
```

> ### Step 4: Running locally

Reviewer: `## Step 4: Run Dagster locally`

Or something along these lines

> In order to run our newly defined Dagster Job we'll need to define a Dagster [Repository](/concepts/repositories-workspaces/repositories). A Dagster Repository is a logical grouping of Jobs and Schedules that provides an entrypoint for viewing and executing Jobs.

Reviewer 2: ```suggestion
In order to run our newly defined Dagster job we'll need to define a Dagster [repository](/concepts/repositories-workspaces/repositories). A Dagster repository is a logical grouping of jobs and schedules that provides an entrypoint for viewing and executing them.
```

> In order to run our newly defined Dagster job we'll need to define a Dagster [repository](/concepts/repositories-workspaces/repositories). A Dagster repository is a logical grouping of jobs and schedules that provides an entrypoint for viewing and executing them.

Reviewer: > To run our newly defined Dagster job, we'll need to define a Dagster [repository](/concepts/repositories-workspaces/repositories). A Dagster repository is a logical grouping of jobs and schedules that provides an entry point for viewing and executing them.

Editing for clarity. With the shift to `Definitions`, we should update this section if everything will work with `Definitions`.

```python file=/integrations/airflow/tutorial_rewrite_dagster.py startafter=start_repo endbefore=end_repo
@repository
def rewrite_repo():
    return [tutorial_job, schedule]
```

> We can now load this file with dagit, the interactive web-ui for dagster

Reviewer: > We can now load this file with Dagit, Dagster's web-based UI:

Additionally, I think it'd be helpful to tell users where they need to run this command.

```bash
dagit -f <your_dagster_file>.py
```

> ## Completed Code Example

Reviewer: > Completed code example

Making the heading use sentence case

> That's it! By now, your code should look like this:

Adding copy to this section to make it (super) clear what it is.

```python file=/integrations/airflow/tutorial_rewrite_dagster.py startafter=start_job endbefore=end_job
@job(tags={"dagster/max_retries": 1, "dag_name": "example"})
def tutorial_job():
    ds = print_date()
    sleep(ds)
    templated(ds)
```

##### Job Level Retries

Job level retries are managed by the [run launcher](/deployment/run-retries) you will need to enable support for them in your dagster.yaml, once you do you can define the retry count on the job.

### Step 3: Defining the schedule

Finally we'll add a schedule

```python file=/integrations/airflow/tutorial_rewrite_dagster.py startafter=start_schedule endbefore=end_schedule
schedule = ScheduleDefinition(job=tutorial_job, cron_schedule="@daily")
```

### Step 4: Running locally

In order to run our newly defined Dagster Job we'll need to define a Dagster [Repository](/concepts/repositories-workspaces/repositories).

```python file=/integrations/airflow/tutorial_rewrite_dagster.py startafter=start_repo endbefore=end_repo
@repository
def rewrite_repo():
    return [tutorial_job, schedule]
```

We can now load this file with dagit, the interactive web-ui for dagster

```bash
dagit -f <your_dagster_file>.py
```

> for reference here's what the file should look like

Reviewer 2: This should be a separate section "## Completed Code Example", similar to what we have in Snowflake: https://docs.dagster.io/integrations/snowflake

---
title: "Learning Dagster from Airlfow"
description: How to get started with Dagster from an Airflow background
---

# Learning Dagster from Airflow

Airflow users often come to Dagster for a variety of reasons. Some want to take advantage of our great story around testability. Others want to deploy to an orchestrator that is container native and has better support for multiple Python environments. And still others are attracted to the promise of a better abstraction: software-defined assets.

In this tutorial, we'll help you make the switch from Airflow to Dagster. Map the concepts you're familiar with in Airflow to Dagster and provide some examples of the work you did in Airflow can be done in Dagster.

By the end of this tutorial, you will:

- Understand how concepts in Airflow map onto Dagster

> - Understand how to author a Dagster Pipeline similiar to a Scheduled Airflow DAG

Reviewer 2: There will be a single point of this tutorial: we will define the DAG from the [Airflow tutorial](https://airflow.apache.org/docs/apache-airflow/stable/tutorial/fundamentals.html) as a Dagster job.

---

> ## Terminology

Reviewer 2: Move all the concepts to a separate page. Only mention the subset of concepts that are used in this tutorial.

| Datasets              | [Software-Defined Assets](/concepts/assets/software-defined-assets)                                                                                                                                                                                                                             | SDAs are much more powerful and mature than datasets, include support for things like partitioning                                                                                                |
| Connections/Variables | [Run Config](/concepts/configuration/config-schema), [Configured APIs](/concepts/configuration/configured) and [Environment Variables](https://docs.dagster.io/dagster-cloud/developing-testing/environment-variables-and-secrets#dagster-cloud-environment-variables-and-secrets) (cloud only) |                                                                                                                                                                                                   |

## Comparing an Airflow Dag to Dagster

> In this guide we will rewrite an Airflow Dag as a Dagster [Job](/concepts/ops-jobs-graphs/jobs).

Reviewer 2: ```suggestion
In this guide we will rewrite an [Airflow DAG](https://airflow.apache.org/docs/apache-airflow/stable/tutorial/fundamentals.html) as a Dagster [Job](/concepts/ops-jobs-graphs/jobs).
```

        retries=3,
    )
    t1.doc_md = dedent(
        """\
    #### Task Documentation
    You can document your task using the attributes `doc_md` (markdown),
    `doc` (plain text), `doc_rst`, `doc_json`, `doc_yaml` which gets
    rendered in the UI's Task Instance Details page.
    ![img](http://montcs.bloomu.edu/~bobmon/Semesters/2012-01/491/import%20soul.png)
    **Image Credit:** Randall Munroe, [XKCD](https://xkcd.com/license.html)
    """
    )

    dag.doc_md = __doc__  # providing that you have a docstring at the beginning of the DAG; OR
    dag.doc_md = """
    This is a documentation placed anywhere
    """  # otherwise, type it like this
    templated_command = dedent(
        """
    {% for i in range(5) %}
        echo "{{ ds }}"
        echo "{{ macros.ds_add(ds, 7)}}"
    {% endfor %}
    """
    )

    t3 = BashOperator(
        task_id="templated",
        depends_on_past=False,
        bash_command=templated_command,
    )

    t1 >> [t2, t3]
```

In order to rewrite this DAG in Dagster we will break it down into three parts

1. The graph of compute
2. The definition of the schedule
3. The configuration of that graph and schedule

> ### Defining the graph

Reviewer 2: Let's use the steps that you mentioned above. I think we can do configuration as we go.

1. Define the computations: "the `op`s" - in Airflow, the operators
2. Define the graph: the `job` - in Airflow, the DAG
3. Define the schedule - In Airflow, the schedule (how simple!)

```suggestion
### Step 1: Define the computations
```

A [Job](/concepts/ops-jobs-graphs/jobs) is made up of a [Graph](/concepts/ops-jobs-graphs/graphs) of [Ops](/concepts/ops-jobs-graphs/ops). If you've used the Airflow taskflow api this will feel familiar, With ops you will be focused on writing a graph with python functions as the nodes with the data dependencies between those as edges.

After rewriting the core compute of the tutorial Airflow Dag in Dagster we're left with:

> ```python

Reviewer 2: This should be interpolated from a Python file

import time
from datetime import date, timedelta

from dagster import job, op

> @op
> def print_date():
>     dt = date.today()
>     print(dt)
>     return dt

Reviewer 2: We should be very precise in the mapping here. Include the docstrings, retries, use `shutil` rather than `datetime` to emulate the `BashOperator`, etc.

```suggestion
@op(name="print_date")
def t1():
    # use shutil
```

@op
def sleep(dt: date):
    time.sleep(5)

@op
def templated(dt: date):
    for i in range(5):
        print(dt)
        print(dt - timedelta(days=7))

@job
def tutorial_job():
    dt = print_date()
    sleep(dt)

>     templated(dt)

Reviewer 2: This should be mentioned in step 2: Defining the graph

@repository
def rewrite_repo():

>     return [tutorial_job]

Reviewer 2: not mentioned in the airflow tutorial, we should not expose this yet. Maybe mention this in Step 4: Running locally?

I think this can be in a separate PR. Steps 4-7 is what's covered in the Testing section in the Airflow quickstart.

```python
@op(
    retry_policy=RetryPolicy(
        max_retries=3
    )
)
def sleep(dt: date):
    time.sleep(5)
```

#### Tagging and Metadata

Dagster has a [tagging and metadata system](/concepts/ops-jobs-graphs/metadata-tags) that is quite a bit more powerful and integrated with the UI than what exists in Airflow.

```python
@job(tags={"dagster/max_retries": 1, "dag_name": "example"})
def tutorial_job():
    dt = print_date()
    sleep(dt)
    templated(dt)
```

### The final product

```python
import time
from datetime import date, timedelta

from dagster import RetryPolicy, ScheduleDefinition, job, op, repository

@op
def print_date():
    dt = date.today()
    print(dt)
    return dt

@op(
    retry_policy=RetryPolicy(
        max_retries=3
    )
)
def sleep(dt: date):
    time.sleep(5)

@op
def templated(dt: date):
    for i in range(5):
        print(dt)
        print(dt - timedelta(days=7))

@job(tags={"dagster/max_retries": 1, "dag_name": "example"})
def tutorial_job():
    dt = print_date()
    sleep(dt)
    templated(dt)

schedule = ScheduleDefinition(job=tutorial_job, cron_schedule="@daily")

@repository
def rewrite_repo():
    return [tutorial_job, schedule]

> ```

Reviewer 2: I think all of this should be done within the steps, and should not be mentioned separately. Otherwise, we are mixing the purposes of this guide with concept mapping.

Reviewer 3: Of these two options, I would vote to include the concepts in the tutorial, or at least to include retries. Retries are a critical part of Airflow that Prefect also makes a big deal about. The mapping to dagster concepts is not straightforward (op retries vs job retries) so it'd be nice to elaborate in the tutorial.

Reviewer 2: I think to have a baseline for this tutorial, we should only focus mapping the Airflow tutorial to its Dagster equivalent.

Retries at the operator and DAG level are mentioned in the Airflow tutorial, so we should definitely mention them in our tutorial.

But my suggestion is not to go too into depth in the tutorial for what the concept is. We should just link out the Concept Mapping guide, which will explain the parity between Airflow's retries and our retries. I'm thinking at the end of each step in our tutorial, we link a "See also: `retries`, `ops`, ...". There, the user can see the in-depth comparison between the specific concepts.

Otherwise, we risk information dumping so much content onto the new user.
